[DIA] Km and Cox functions

vanessavieira opened the pull request:
* Added DML script for Km and Cox based on existing ones (improving script, removing read/write statements, etc)
* Added functions to Builtin.java
* Added set-up for test class for both Km and Cox (still missing tests)

vanessavieira commented:
Finished implementation + testing. Unfortunately I have some issues with returning the values from the cox script, I keep having a SingularMatrixException but I do not know how to fix it. I have tried many things but it did not work so far. Any help from the reviewers is much appreciated.

Baunsgaard reviewed (commented):
> Finished implementation + testing. Unfortunately I have some issues with returning the values from the cox script, I keep having a SingularMatrixException but I do not know how to fix it. I have tried many things but it did not work so far. Any help from the reviewers is much appreciated.

I see there is an issue, i will take a closer look later if i can spot the solution it is most likely something in the declaration of the cox script.

vanessavieira commented on src/test/java/org/apache/sysds/test/functions/builtin/BuiltinCoxTest.java:
```diff
@@ -0,0 +1,58 @@
+				"v=" + shapeWeibull, "p=" + prob, "m=" + numFeatures, "sp=" + sparsity,
+				"alpha=" + alpha, "tol=" + tol, "moi=" + moi, "mii=" + mii};
+
+		runTest(true, false, null, -1);
```
I see and I agree I should be testing the output. The problem I faced with testing is how to know the output I am expecting. First I needed to know how the input should look like and for that, I used parts of the data generation for survival analysis script (https://github.com/apache/systemds/blob/master/scripts/datagen/genRandData4SurvAnalysis.dml) but with that I don't really have a way of knowing what should be the output. Do you think I should create an R script that accepts the same input and then checks if the output returned is the same? I'm not sure how much trouble that would be or if there's already both algorithms implemented in R

Baunsgaard replied:
it is a challenge to test the algorithms, if there is an equivalent algorithm in R then that would be fine to compare with. But best would be if we could test it ourselves. My suggestion is to keep it simple, and try to think of the algorithm as a black box. If you know what it does, then try to deconstruct what behavior is expected based on as simple an input as possible. and then expand from there.

I don't expect you to test in all edge cases, just that we know it works on commonly expected inputs.

vanessavieira replied:
@Baunsgaard could you send me your email so I can maybe describe better some questions I still have regarding this? Unfortunately both algorithms are not simple enough so I know exactly the output in those simple cases. I feel extremely lost. Even in simple cases I am not sure how the formatting should be for the timestamp generation, for example. And by using the script for data generation I have no way of knowing exactly what the input is to know what the output should be as well. If you could send me your email we could maybe talk better to what I am expected to do here.

Baunsgaard replied:
baunsgaard@tugraz.at

vanessavieira commented on src/test/scripts/functions/builtin/cox.dml:
```diff
@@ -0,0 +1,69 @@
+
+v = $v;
+# generate feature matrix
+X_t = rand (rows = num_records, cols = num_features, min = 1, max = 5, pdf = "uniform", sparsity = sparsity);
```
I tried that in the beginning but I was a bit unsure if what I coded in Java was really reflecting the data generation of this script (https://github.com/apache/systemds/blob/master/scripts/datagen/genRandData4SurvAnalysis.dml). I could try that again though.

vanessavieira commented on scripts/builtin/km.dml:
```diff
@@ -0,0 +1,656 @@
+# ---------------------------------------------------------------------------------------------
+# X       String   ---          Location to read the input matrix X containing the survival data: 
+#								timestamps, whether event occurred (1) or data is censored (0), and a number of factors (categorical features) 
+#								for grouping and/or stratifying 
```
Oh, I must have missed that information. Will add space instead of tabs. Thanks :)

vanessavieira commented:
* Fixed identation issues for cox and km
* Added a seed to the random data generator for running the script
* For cox.dml I still face the SingleMatrixException. I did fix the return for S and some of T and for some of M. The parts of the script that are commented with # indicates the line where when it's run, I have the exception. I still could not figure it out why.

Shafaq-Siddiqi commented:
Hi,
Thank you for your contribution. The PR looks in a good shape. Just one minor task if you could please add the R scripts and compare the functions with equivalent R functions. You can have a look into other built-in i.e., winsorize, to see how we are comparing them against R functions.

vanessavieira commented:
> Hi,
> Thank you for your contribution. The PR looks in a good shape. Just one minor task if you could please add the R scripts and compare the functions with equivalent R functions. You can have a look into other built-in i.e., winsorize, to see how we are comparing them against R functions.

Could you help me figure it out how I should have the input in the R script? You gave me the idea to generate the input data within the test/*.dml script and so I am not really sure how I should do with the R script

Shafaq-Siddiqi commented:
> > Hi,
> > Thank you for your contribution. The PR looks in a good shape. Just one minor task if you could please add the R scripts and compare the functions with equivalent R functions. You can have a look into other built-in i.e., winsorize, to see how we are comparing them against R functions.
> 
> Could you help me figure it out how I should have the input in the R script? You gave me the idea to generate the input data within the test/*.dml script and so I am not really sure how I should do with the R script

You need to search for the Km() and cox() functions in the R library then create a cox.R file and call the R implementation of cox() on the same dataset that you are using in cox.dml file and write the results. In the BuiltinCox.java file run the R script and read both R and DML output and compare them. Do the same for km. You can refer to winsorize.R script to see how the inputs are read and written.

mboehm7 commented:
thanks for your effort @vanessavieira - I'll take a look into the remaining issues and merge this next week.

mboehm7 commented:
ok thanks @vanessavieira - I now did some basic cleanups of the scripts, and removed the empty comparison files. Could you please associate your github account with the email address you used for the commits. Thanks.

I'll do the output comparison in the tests and related fixes in the builtin functions as a follow up project.

vanessavieira commented:
> ok thanks @vanessavieira - I now did some basic cleanups of the scripts, and removed the empty comparison files. Could you please associate your github account with the email address you used for the commits. Thanks.
> 
> I'll do the output comparison in the tests and related fixes in the builtin functions as a follow up project.

All right, thank you! From what I see in the settings my github is already associated with my primary email address (vieira.vanessasoares@gmail.com).

mboehm7 commented:
well, your commits used your tugraz email address and thus github is unable to link the commit to your handle - you can associate multiple emails addresses with your github account.

vanessavieira commented:
Aaa, I see. I also had another institutional email from my bachelor's university (vsv@ic.ufal.br) so I removed that one and added my tugraz email (vanessa.soaresvieira@student.tugraz.at) to my github account. Hope it's all good now?

mboehm7 commented:
awesome - thanks.
